shell/common/shell.cc

```diff
@@ -736,47 +736,21 @@ void Shell::OnPlatformViewDestroyed() {
     fml::TaskRunner::RunNowOrPostTask(io_task_runner, io_task);
   };
 
-  // The normal flow executed by this method is that the platform thread is
-  // starting the sequence and waiting on the latch. Later the UI thread posts
-  // raster_task to the raster thread triggers signaling the latch(on the IO
-  // thread). If the raster and the platform threads are the same this results
-  // in a deadlock as the raster_task will never be posted to the plaform/raster
-  // thread that is blocked on a latch.  To avoid the described deadlock, if the
-  // raster and the platform threads are the same, should_post_raster_task will
-  // be false, and then instead of posting a task to the raster thread, the ui
-  // thread just signals the latch and the platform/raster thread follows with
-  // executing raster_task.
-  bool should_post_raster_task = task_runners_.GetRasterTaskRunner() !=
-                                 task_runners_.GetPlatformTaskRunner();
-
-  auto ui_task = [engine = engine_->GetWeakPtr(),
-                  raster_task_runner = task_runners_.GetRasterTaskRunner(),
-                  raster_task, should_post_raster_task, &latch]() {
+  auto ui_task = [engine = engine_->GetWeakPtr(), &latch]() {
     if (engine) {
       engine->OnOutputSurfaceDestroyed();
     }
-    // Step 1: Next, tell the raster thread that its rasterizer should suspend
-    // access to the underlying surface.
-    if (should_post_raster_task) {
```

Comment:
@cyanglaz I was thinking about this. Sorry for all these questions:

1. Is it possible that `should_post_raster_task` is `true` at this point even though it should be `false`? 
2. Why isn't this checking a global shared state (such as `RasterThreadMerger::IsMerged()`)  instead of a local state?
3. Could this be the reason why we need to merge the threads to fix the race condition?

Reply:
1. `should_post_raster_task ` was only exists for static thread merging. This value is always `true` for dynamic thread merging.
2. I'm not sure why it was a local var, but global state would also work.
3. This is the reason why we needed to merge threads here. Because thread merging happens on raster thread and this is running on platform thread, we cannot know if threads are merged.

Reply:
If the platform thread is blocked at this point waiting on the latch, and the raster queue is running on the platform thread, can we release the latch and indicate that the raster should be teared down right after the latch is awaited?